File: app/forms/signup_form.py
from flask_wtf import FlaskForm
from wtforms import StringField
from wtforms.validators import DataRequired, Email, ValidationError, Length
from app.models import User


def useremail_exists(form, field):
    # Checking if user exists
    email = field.data
    user = User.query.filter(User.email == email).first()
    if user:
        raise ValidationError('Email address is already in use.')


def valid_email(form, field):
    email = field.data
    if '@' not in email:
        raise ValidationError("Invalid email address.")


def username_exists(form, field):
    # Checking if username is already in use
    username = field.data
    user = User.query.filter(User.username == username).first()
    if user:
        raise ValidationError('Username is already in use.')


# def valid_password(form, field):
#     password = field.data
#     if len(password) < 6:
#         raise ValidationError("Passwords must be at least 6 characters.")


class SignUpForm(FlaskForm):
    firstname = StringField(
        'firstname', validators=[DataRequired('First name is required.'), Length(max=50, message='First name cannot exceed 50 characters.')])
    lastname = StringField(
        'lastname', validators=[DataRequired('Last name is required.'), Length(max=50, message='Last name cannot exceed 50 characters.')])
    username = StringField(
        'username', validators=[DataRequired('User name is required.'), username_exists, Length(max=50, message='User name cannot exceed 50 characters.')])
    email = StringField(
        'email', validators=[DataRequired('Enter your email'), useremail_exists, valid_email, Length(max=255, message='Email cannot exceed 255 characters.')])
    password = StringField(
        'password', validators=[DataRequired('Enter your password'), Length(min=6, max=255, message='Password should be between 6 to 255 characters.')])

File: react-app/src/components/auth/SignUpForm.js
import React, { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux'
import { Redirect, NavLink } from 'react-router-dom';
import { signUp } from '../../store/session';
import mzonelogo from '../../img/mzonelogowhite.png'
import './LoginForm.css';

const SignUpForm = () => {
  const [errors, setErrors] = useState([]);
  const [username, setUsername] = useState('');
  const [email, setEmail] = useState('');
  const [firstname, setFirstname] = useState('');
  const [lastname, setLastname] = useState('');
  const [password, setPassword] = useState('');
  const [repeatPassword, setRepeatPassword] = useState('');
  const user = useSelector(state => state.session.user);
  const dispatch = useDispatch();

  const onSignUp = async (e) => {
    e.preventDefault();
    if (password === repeatPassword) {
      const data = await dispatch(signUp(firstname, lastname, username, email, password));
      if (data) {
        setErrors(data)
      }
    }
  };

  const updateFirstname = (e) => {
    setFirstname(e.target.value);
  };

  const updateLastname = (e) => {
    setLastname(e.target.value);
  };

  const updateUsername = (e) => {
    setUsername(e.target.value);
  };

  const updateEmail = (e) => {
    setEmail(e.target.value);
  };

  const updatePassword = (e) => {
    setPassword(e.target.value);
  };

  const updateRepeatPassword = (e) => {
    setRepeatPassword(e.target.value);
  };

  if (user) {
    return <Redirect to='/' />;
  }

  return (
    <div className='login-page-container'>
      <div className='login-page-container-inner'>
        <NavLink to={`/`}>
          <img src={mzonelogo} className='login-page-logo' alt='mzone' />
        </NavLink>
        <div className='login-page-form-container'>
          <form onSubmit={onSignUp}>
            <div className='login-page-form-title'>Create account</div>
            <div>
              {errors.map((error, ind) => (
                <div className='login-page-form-errors-container'>
                  <div className='login-page-form-errors1'>!</div>
                  <div key={ind}>{error}</div>
                </div>
              ))}
            </div>
            <div className='login-page-form-email-container'>
              <label className='login-page-form-email1'>First Name</label>
              <input
                type='text'
                name='firstname'
                onChange={updateFirstname}
                value={firstname}
              ></input>
            </div>
            <div className='login-page-form-email-container'>
              <label className='login-page-form-email1'>Last Name</label>
              <input
                type='text'
                name='lastname'
                onChange={updateLastname}
                value={lastname}
              ></input>
            </div>
            <div className='login-page-form-email-container'>
              <label className='login-page-form-email1'>User Name</label>
              <input
                type='text'
                name='username'
                onChange={updateUsername}
                value={username}
              ></input>
            </div>
            <div className='login-page-form-email-container'>
              <label className='login-page-form-email1'>Email</label>
              <input
                type='text'
                name='email'
                onChange={updateEmail}
                value={email}
              ></input>
            </div>
            <div className='login-page-form-email-container'>
              <label className='login-page-form-email1'>Password</label>
              <input
                type='password'
                name='password'
                placeholder='At least 6 characters'
                onChange={updatePassword}
                value={password}
              ></input>
              <div className='signup-page-form-password3'>
                <div className='signup-page-form-password3icon'>i</div>
                <div className='signup-page-form-password3word'> Passwords must be at least 6 characters.</div>
              </div>
            </div>
            <div className='login-page-form-email-container'>
              <label className='login-page-form-email1'>Re-enter password</label>
              <input
                type='password'
                name='repeat_password'
                onChange={updateRepeatPassword}
                value={repeatPassword}
              // required={true}
              ></input>
            </div>
            <button type='submit'>Sign Up</button>
            <div className='login-page-form-note'>
              By continuing, you agree to Mzone's Conditions of Use and Privacy Notice.
            </div>
          </form>
          <div className='signup-page-line1'></div>
          <div className='signup-page-loginacct'>
            <div className='signup-page-haveacct'>
              Already have an account?
            </div>
            <NavLink to={`/login`}>Sign in</NavLink>
          </div>


        </div>
      </div>
    </div>
  );
};

export default SignUpForm;
